### DISCUSSION TITLE
superbrewer3000.notation exception when opening in development environment

### DISCUSSION THREAD

**Discussion by e-lynch-theia** [eclipse-emfcloud/emfcloud#211] [Coffee Editor] [answered]
Hi,
I am running the coffee editor through development environment (eclipse and VSCode) and I am getting an exception when opening the superbrewer3000.notation diagram editor.  
Cannot cast org.eclipse.glsp.server.emf.model.notation.impl.DiagramImpl to org.eclipse.emfcloud.coffee.Machine
java.lang.ClassCastException: Cannot cast org.eclipse.glsp.server.emf.model.notation.impl.DiagramImpl to org.eclipse.emfcloud.coffee.Machine
	at java.base/java.lang.Class.cast(Class.java:3889)
	at org.eclipse.emfcloud.coffee.workflow.glsp.server.gmodel.WorkflowGModelFactory.fillRootElement(WorkflowGModelFactory.java:61)
	at org.eclipse.emfcloud.modelserver.glsp.notation.integration.EMSNotationGModelFactory.fillRootElement(EMSNotationGModelFactory.java:35)
	at org.eclipse.emfcloud.modelserver.glsp.EMSGModelFactory.createGModel(EMSGModelFactory.java:37)

Can someone please direct me as to how to fix this problem please?
Thank you
Evelyn

**Comment by ndoschek** [comment]
Hi @e-lynch-theia!

Thanks for your interest in the coffee-editor :tada: !
Hm that sounds off. If I read the error message correctly, it tries to cast the notation model to the semantic model, which should not happen in any case.
- Which version are you using?
- Did you build all the components via the docker container or locally? And if locally, also [according to the prerequisites](https://github.com/eclipse-emfcloud/coffee-editor#prerequisites)?
- Do you have any changes on the files in the opened workspace perhaps?

You're probably already aware of that, but just for the sake of completeness, you could also try out the coffee-editor online: https://try.theia-cloud.io/?appdef=coffee-editor

and just a heads up:
Most of us will be ooo for the rest of the week (due to a public holiday and a bridge day), if there any follow-up questions we will come back to you next week :)

Best, Nina

**Comment by e-lynch-theia** [comment]
Hi Nina,
Thank you for swift response.   Here are the answers to your questions :


  *   Which version are you using?  I pulled from the latest https://github.com/eclipse-emfcloud/coffee-editor.git
  *   Did you build all the components via the docker container or locally? And if locally, also according to the prerequisites<https://github.com/eclipse-emfcloud/coffee-editor#prerequisites>?  I am building the coffee editor locally and I did follow the prerequisites
  *   Do you have any changes on the files in the opened workspace perhaps? no changes in workspace files

I have looked at the editor online but now I need to get this going on my development environment. I want to see how/if we might migrate to this technology - in the future. We are already an eclipse rcp desktop application and looking to see if this would be a good fit for us.

Any ideas on where this has gone wrong?
Thanks
Evelyn


Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by ndoschek** [reply]
Hi again @e-lynch-theia,

Thanks for your response!
I had a look and cloned a fresh version of the coffee-editor on my machine.
I noticed some inconsistencies after the last update and pushed an update (mainly regarding an inconsistent use of a dependency on the frontend and missing plugins on the server side for the WorkflowAnalyzer launch config). However, those shouldn't have caused your mentioned server error.

Based on this, I tested the editor in development mode and with the embedded servers and unfortunately, I could not reproduce your issue, but I list my steps below if I run the components in "developer mode" to be able to debug them, just in case:

1. build all components with `yarn build` in the root directory
2. open an Eclipse workspace and import all existing projects as Maven projects from `coffee-editor/backend`
3. open `coffee-editor/backend/releng/org.eclipse.emfcloud.coffee.target/org.eclipse.emfcloud.coffee.target.target`, resolve target platform and set as active target platform
4. launch all three launch configs in `coffee-editor/backend/releng/org.eclipse.emfcloud.coffee.product/launch` (that launches 4 processes, i.e. the group of the model server and the glsp server as well as the analyzer server and the DSL server)
5. open the `coffee-editor/client` directory in VSCode
6. launch the client backend via the `Launch Browser Backend [DEBUG] (External Servers)` config in the debug view of VSCode
7. open `http://localhost:3000` in the browser (or launch the `Launch Browser Frontend` config via the debug view)

Could you please provide any additional logs, e.g. from the client application (back- and frontend).

Thanks and best regards,
Nina

**Comment by e-lynch-theia** [comment]
Hi Nina,
I have cloned the coffee editor again and completed the steps as stated above.  Unfortunately, I am still having the same problems.  I am seeing different problems depending on whether I am running in development mode or otherwise.
I will attach some logs so maybe you will be able to see my problems?    
Log for "yarn build"
[yarn-build.log](https://github.com/eclipse-emfcloud/emfcloud/files/11787192/yarn-build.log)
Log for "yarn start" - Since I am running on Windows I did have to update the start line in the package.json to use "set" instead of "export"
[yarn-start.log](https://github.com/eclipse-emfcloud/emfcloud/files/11787232/yarn-start.log)
One of the main errors I notice here is that in the logs - this location doesn't exist:  C:\Theia_coffee_editor_integrated\coffee-editor\client\node_modules\coffee-servers\servers\org.eclipse.emfcloud.coffee.workflow.glsp.server-0.8.0-SNAPSHOT-glsp.jar???


Errors when running in development mode-
[LaunchBrowserBackend.log](https://github.com/eclipse-emfcloud/emfcloud/files/11787287/LaunchBrowserBackend.log)
[LaunchBrowserFrontend.log](https://github.com/eclipse-emfcloud/emfcloud/files/11787288/LaunchBrowserFrontend.log)
[WorkflowGLSPServer.log](https://github.com/eclipse-emfcloud/emfcloud/files/11787289/WorkflowGLSPServer.log)
[CoffeeModelServer.log](https://github.com/eclipse-emfcloud/emfcloud/files/11787290/CoffeeModelServer.log)

Sorry about all the logs.  Hope you see something? 
Thanks
Evelyn

**Reply by ndoschek** [reply]
Hi Evelyn!

Thanks for providing the logs!
As far as I can see, the built server products were copied over correctly, then it's weird that it cannot find them via `yarn start`...
I will test it on a Windows VM tomorrow and will check if I can spot anything.

Best, Nina

**Reply by ndoschek** [reply]
Hi Evelyn, I setup my Windows VM and could reproduce the issue. It is most likely an issue with the URI handling, that has some troubles interpreting Windows paths. I will look into this and hope to provide a quick fix soon.
I'll ping you again once its resolved!
Best, Nina

**Comment by e-lynch-theia** [comment]
Great Nina that you see the same problem. Looking forward to getting the solution soon. Thank you.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by ndoschek** [reply]
Hi again Evelyn,

I pushed a branch that fixes the initial problem with the diagram editor.
Please see here: https://github.com/eclipse-emfcloud/coffee-editor/tree/windows-compatibility

I also noticed that some other features (i.e. the Model Analysis, Java Code Generator and CPP Code Generator) fail on Windows (due to the same root problem, that Windows paths are not correctly interpreted).
I added a quickfix workaround in that branch as well, but we will take a deeper look to solve it properly.

But in the meantime, on this branch you should be able to run the coffee-editor on Windows and test all the features listed in the Getting Started guide. Hope it works for you as well, and please let me know if there are any issues.

Best regards and have a nice weekend,
Nina

**Comment by e-lynch-theia** [comment]
Hi Nina
That is working now, thank you.  

There still seems to be a problem with the Model Analysis.  I am getting the following error : 
Error: No connection to model analysis server
    at C:\Theia_Coffee_editor_integrated\coffee-editor\client\coffee-servers\lib\node\workflow-analysis-server.js:111:24
    at new Promise (<anonymous>)
    at WorkflowAnalysisServer.analyze (C:\Theia_Coffee_editor_integrated\coffee-editor\client\coffee-servers\lib\node\workflow-analysis-server.js:106:16)
    at JsonRpcProxyFactory.onRequest (C:\Theia_Coffee_editor_integrated\coffee-editor\client\node_modules\@theia\core\lib\common\messaging\proxy-factory.js:132:49)
    at RpcProtocol.requestHandler (C:\Theia_Coffee_editor_integrated\coffee-editor\client\node_modules\@theia\core\lib\common\messaging\proxy-factory.js:113:84)
    at RpcProtocol.handleRequest (C:\Theia_Coffee_editor_integrated\coffee-editor\client\node_modules\@theia\core\lib\common\message-rpc\rpc-protocol.js:167:39)
    at RpcProtocol.handleMessage (C:\Theia_Coffee_editor_integrated\coffee-editor\client\node_modules\@theia\core\lib\common\message-rpc\rpc-protocol.js:64:26)
    at C:\Theia_Coffee_editor_integrated\coffee-editor\client\node_modules\@theia\core\lib\common\message-rpc\rpc-protocol.js:47:66
    at C:\Theia_Coffee_editor_integrated\coffee-editor\client\node_modules\@theia\core\lib\common\event.js:103:69
    at CallbackList.invoke (C:\Theia_Coffee_editor_integrated\coffee-editor\client\node_modules\@theia\core\lib\common\event.js:109:26)

I have all 4 servers started.  Again, using windows development so maybe it is the same issue?
Thank you for the quick response.
Evelyn

**Reply by ndoschek** [reply]
Hi Evelyn, I will have a look on Monday on this!

**Comment by e-lynch-theia** [comment]
No bother Nina, I have what I need for now. Thank you.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by ndoschek** [reply]
Hi again Evelyn,

glad to hear! However I could spot the problem, the connection to the embedded Workflow Analyzer server was erroring, I added a commit on top of the branch, that will fix this issue.
https://github.com/eclipse-emfcloud/coffee-editor/tree/windows-compatibility

Best regards, Nina

**Comment by ndoschek** [comment, marked as answer]
FYI: We merged the compatibility fixes to master in the meantime.
